DDF-04261 Optimize MetacardTypeImpl getAttributeDescriptor

troymohl opened the pull request:
#### What does this PR do?
Optimizes the `MetacardTypeImpl.getAttributeDescriptor` call such that a lazy-initialized map is created the first time it is called, and used for subsequent calls after that to avoid a O(n) operation on each call.

#### Who is reviewing it? 
@zta6 
@leo-sakh 

#### Select relevant component teams: 
@codice/core-apis 

#### Ask 2 committers to review/merge the PR and tag them here.
@bdeining
@pklinef
@rzwiefel 
@clockard 

#### How should this be tested?
Full build with tests.

#### Any background context you want to provide?
The lazy-initialization approach was taken due to the number of places that `descriptors` is directly added to by subclasses. This keeps the changeset small while still providing a performance improvement benefit.

#### What are the relevant tickets?
Fixes: #04261

#### Screenshots

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

leo-sakh commented on catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/MetacardTypeImpl.java:
```diff
@@ -54,6 +55,8 @@
   /** Set of {@link AttributeDescriptor}s */
   protected transient Set<AttributeDescriptor> descriptors = new HashSet<>();
 
+  protected transient HashMap<String, AttributeDescriptor> descriptorNameMap = new HashMap<>();
```
✏️ Since we're only using regular `Map` methods on this reference, why not just leave it as `Map<String, AttributeDescriptor>`?

troymohl replied:
oops. fixed that.

zta6 commented on catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/MetacardTypeImpl.java:
```diff
@@ -157,13 +160,14 @@ public AttributeDescriptor getAttributeDescriptor(String attributeName) {
-    for (AttributeDescriptor descriptor : descriptors) {
-      if (attributeName.equals(descriptor.getName())) {
-        return descriptor;
+
```
✏️ this is meant to be done just once at initialization, right? Perhaps it would be best done in the constructor, or in `validateDescriptors()`. I don't have much preference as to whether it goes here or somewhere else, just sharing a thought.

zta6 commented on catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/MetacardTypeImpl.java:
```diff
@@ -157,13 +160,14 @@ public AttributeDescriptor getAttributeDescriptor(String attributeName) {
-      if (attributeName.equals(descriptor.getName())) {
-        return descriptor;
+
+    if (descriptors.size() != descriptorNameMap.size()) {
```
If the above is the case:
✏️ perhaps you could use a flag variable like `shouldUpdateNameMap`, set that to true every time the set of attribute descriptors is modified, then check the flag here, update the name map and reset the flag if needed. 
That way you wouldn't have to check through to see if all of the attributes are actually the same.

Alternatively, you might be able to add this logic to the end of `addAttributeDescriptors` and refactor the one constructor to use `addAttributeDescriptors` instead of `addAll()`.

troymohl replied:
@coyotesqrl The case could occur, but I don't think it's too likely. Most of the time subclasses are just adding descriptors. 
@zta6 I don't know of any to add a listener to a hashset that fires when the set changes. The problem is subclasses are directly adding descriptors via `descriptors.addAll()`

ethantmanns replied:
My opinion would be that it's better to be slow and always correct

troymohl commented:
@coyotesqrl & @ethantmanns - I have updated the PR to change the underlying collection to a Map and refactored dependent code in DDF to use new method for `add` and `addAll`

coyotesqrl commented:
Looks good. 

I've also added the `Impacts Downstream` label as it should be reflected in any downstream, concrete metacard types.

troymohl commented:
build now

rzwiefel commented:
@troymohl do we have metrics on performance before and after this change? I've looked at this before but haven't been able to provide concrete evidence the O(n) finds are cause more slowdown than the overhead added by the set creation for the get attribute descriptors and add all methods?

troymohl commented:
@rzwiefel With the latest change, the original addition of descriptors to the HashSet was also an O(n) operation. In fact under the covers, HashSet is actually backed by a HashMap in the OpenJDK. So as far as creating the HashMap the cost is O(n) which is the same. The `getAttributeDescriptor` method has been changed from an O(n) operation to O(1) which is always faster. 
As far as measured performance, I was seeing about 15-20% faster performance and this method is no longer a hotspot when ingesting 1000s of records over CSW.

troymohl commented:
build now
